lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionStarter.java

```diff
@@ -0,0 +1,204 @@
+            LockLeaseService lockLeaseService, int numberOfTransactions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> result = new ArrayList<>();
+        while (result.size() < numberOfTransactions) {
+            result.addAll(split(lockLeaseService.startTransactions(numberOfTransactions - result.size())));
+        }
+        return result;
+    }
+
+    StartIdentifiedAtlasDbTransactionResponse startIdentifiedAtlasDbTransaction() {
+        try {
+            return autobatcher.apply(null).get();
+        } catch (ExecutionException e) {
+            throw Throwables.throwUncheckedException(e.getCause());
+        } catch (Throwable t) {
+            throw Throwables.throwUncheckedException(t);
+        }
+    }
+
+    Set<LockToken> refreshLockLeases(Set<LockToken> tokens) {
+        Set<LockTokenShare> lockTokenShares = filterLockTokenShares(tokens);
+        Set<LockToken> lockTokens = filterOutTokenShares(tokens);
+
+        Set<LockToken> refreshedTokens = lockLeaseService.refreshLockLeases(Sets.union(
+                reduceForRefresh(lockTokenShares),
+                lockTokens));
+
+        Set<LockToken> resultLockTokenShares = lockTokenShares.stream()
+                .filter(t -> refreshedTokens.contains(t.sharedLockToken()))
+                .collect(Collectors.toSet());
+        Set<LockToken> resultLockTokens = lockTokens.stream()
+                .filter(refreshedTokens::contains)
+                .collect(Collectors.toSet());
+
+        return Sets.union(resultLockTokenShares, resultLockTokens);
+    }
+
+    Set<LockToken> unlock(Set<LockToken> tokens) {
+        Set<LockToken> lockTokens = filterOutTokenShares(tokens);
+
+        Set<LockTokenShare> lockTokenShares = filterLockTokenShares(tokens);
+        Set<LockToken> referencedTokens = lockTokenShares.stream()
+                .map(LockTokenShare::sharedLockToken)
+                .collect(Collectors.toSet());
+
+        Set<LockToken> toUnlock = reduceForUnlock(lockTokenShares);
+        Set<LockToken> toRefresh = Sets.difference(referencedTokens, toUnlock);
+
+        //calling refresh to check validity of tokens that won't be unlocked
+        Set<LockToken> refreshed = refreshLockLeasesInternal(toRefresh);
+        Set<LockToken> unlocked = lockLeaseService.unlock(Sets.union(toUnlock, lockTokens));
+
+        Set<LockToken> resultLockTokenShares = lockTokenShares.stream()
+                .filter(t -> unlocked.contains(t.sharedLockToken()) || refreshed.contains(t.sharedLockToken()))
+                .collect(Collectors.toSet());
+        Set<LockToken> resultLockTokens = lockTokens.stream().filter(unlocked::contains).collect(Collectors.toSet());
+
+        return Sets.union(resultLockTokenShares, resultLockTokens);
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    private Set<LockToken> refreshLockLeasesInternal(Set<LockToken> toRefresh) {
+        if (!toRefresh.isEmpty()) {
```

Comment:
Can we push these checks down to `LockLeaseService` to simplify the code here? `LockLeaseService#refreshTokens` has an analogous shortcut already.